test/ng/directive/inputSpec.js

```diff
@@ -2621,154 +2621,171 @@ describe('input', function() {
+          $rootScope.$apply('step = 3');
+          expect(inputElm.val()).toBe('15');
+          expect(inputElm).toBeInvalid();
+          expect(ngModel.$error.step).toBe(true);
+          expect($rootScope.value).toBeUndefined();
 
-        helper.changeInputValueTo('10');
-        expect(inputElm).toBeValid();
-        expect($rootScope.value).toBe(10);
+          helper.changeInputValueTo('8');
+          expect(inputElm).toBeValid();
+          expect($rootScope.value).toBe(8);
 
-        // Step changes, but value matches
-        $rootScope.$apply('step = 5');
-        expect(inputElm.val()).toBe('10');
-        expect(inputElm).toBeValid();
-        expect($rootScope.value).toBe(10);
-        expect($rootScope.form.alias.$error.step).toBeFalsy();
+          $rootScope.$apply('min = 10; step = 20');
+          helper.changeInputValueTo('30');
+          expect(inputElm.val()).toBe('30');
+          expect(inputElm).toBeValid();
+          expect($rootScope.value).toBe(30);
 
-        // Step changes, value does not match
-        $rootScope.$apply('step = 6');
-        expect(inputElm).toBeInvalid();
-        expect($rootScope.value).toBeUndefined();
-        expect(inputElm.val()).toBe('10');
-        expect($rootScope.form.alias.$error.step).toBeTruthy();
+          $rootScope.$apply('min = 5');
+          expect(inputElm.val()).toBe('30');
+          expect(inputElm).toBeInvalid();
+          expect(ngModel.$error.step).toBe(true);
+          expect($rootScope.value).toBeUndefined();
 
-        // null = valid
-        $rootScope.$apply('step = null');
-        expect(inputElm).toBeValid();
-        expect($rootScope.value).toBe(10);
-        expect(inputElm.val()).toBe('10');
-        expect($rootScope.form.alias.$error.step).toBeFalsy();
+          $rootScope.$apply('step = 0.00000001');
+          expect(inputElm.val()).toBe('30');
+          expect(inputElm).toBeValid();
+          expect($rootScope.value).toBe(30);
 
-        // Step val as string
-        $rootScope.$apply('step = "7"');
-        expect(inputElm).toBeInvalid();
-        expect($rootScope.value).toBeUndefined();
-        expect(inputElm.val()).toBe('10');
-        expect($rootScope.form.alias.$error.step).toBeTruthy();
+          // 0.3 - 0.2 === 0.09999999999999998
+          $rootScope.$apply('min = 0.2; step = (0.3 - 0.2)');
+          helper.changeInputValueTo('0.3');
+          expect(inputElm.val()).toBe('0.3');
+          expect(inputElm).toBeInvalid();
+          expect(ngModel.$error.step).toBe(true);
+          expect($rootScope.value).toBeUndefined();
+        });
 
-        // unparsable string is ignored
-        $rootScope.$apply('step = "abc"');
-        expect(inputElm).toBeValid();
-        expect($rootScope.value).toBe(10);
-        expect(inputElm.val()).toBe('10');
-        expect($rootScope.form.alias.$error.step).toBeFalsy();
+        it('should correctly validate even in cases where `(x*y % x !== 0)`', function() {
```

Comment:
It short of is, but `value % step === 0` is the obvious way to implement it and it is the exact implementation that leads into errors. I mean we wouldn't have this issues if we had implemented it differently, but we want to explicitly test that the obvious implementation falls short.

That said, I am always open to suggestions :wink: